src/RevExt.cc

```diff
@@ -59,18 +29,27 @@ bool RevExt::Execute( unsigned Inst, const RevInst& payload, uint16_t HartID, Re
     return false;
   }
 
-  // execute the instruction
   bool ret = false;
-
-  if( payload.rm == FRMode::None ) {
-    // If the instruction has no FRMode, we do not need to save and restore it
+  if( !payload.raisefpe ) {
+    // If the instruction cannot raise FP exceptions, don't mess with the FP
+    // environment. No RISC-V FP instructions which cannot raise FP exceptions
+    // depend on the FP rounding mode, i.e. depending on rounding mode implies
+    // that the instruction can raise FP exceptions.
     ret = func( feature, regFile, mem, payload );
   } else {
-    // saved_fenv represents a saved FP environment, which, when destroyed,
-    // restores the original FP environment. We execute the function in the
-    // modified FP environment on the host, then restore the FP environment.
-    auto saved_fenv = SetFPEnv( Inst, payload, HartID, regFile );
-    ret             = func( feature, regFile, mem, payload );
+    // saved_fenv represents a saved FP environment on the host, which, when
+    // destroyed, restores the original FP host environment. We execute the
+    // instruction in a default FP environment on the host with all FP
+    // exceptions cleared and the rounding mode set according to the encoding
+    // and frm register. The FP exceptions which occur on the host are stored
+    // in FCSR when saved_fenv is destroyed.
+    RevFenv saved_fenv( regFile, payload.rm, output );
+
+    // Execute the instruction
+    ret = func( feature, regFile, mem, payload );
+
+    // Fall-through destroys saved_fenv, setting the FCSR's fflags and
+    // restoring the host's FP environment
```

Comment:
**This is the main mechanism for emulating FP rounding modes and exceptions on the host machine.**